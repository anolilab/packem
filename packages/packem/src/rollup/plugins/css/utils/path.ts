import { join, relative, resolve } from "@visulima/path";

export const isAbsolutePath = (path: string): boolean => /^(?:\/|(?:[A-Z]:)?[/\\|])/i.test(path);

export const isRelativePath = (path: string): boolean => /^\.?\.[/\\]/.test(path);

export const normalizePath = (...paths: string[]): string => {
    const f = join(...paths).replaceAll("\\", "/");

    if (/^\.[/\\]/.test(paths[0] as string)) {
        return `./${f}`;
    }

    return f;
};

export const resolvePath = (...paths: string[]): string => normalizePath(resolve(...paths));

export const relativePath = (from: string, to: string): string => normalizePath(relative(from, to));

export const humanlizePath = (file: string): string => relativePath(process.cwd(), file);
