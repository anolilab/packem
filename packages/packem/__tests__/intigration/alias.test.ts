import { rm } from "node:fs/promises";
import { resolve } from "node:path";

import { readFileSync, writeFileSync } from "@visulima/fs";
import { temporaryDirectory } from "tempy";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

import { createPackageJson, createPackemConfig, createTsConfig, execPackemSync, streamToString } from "../helpers";

describe("packem alias", () => {
    let temporaryDirectoryPath: string;

    beforeEach(async () => {
        temporaryDirectoryPath = temporaryDirectory();
    });

    afterEach(async () => {
        await rm(temporaryDirectoryPath, { recursive: true });
    });

    it("should not trigger a warning if alias option is used", async () => {
        expect.assertions(4);

        writeFileSync(
            `${temporaryDirectoryPath}/src/index.ts`,
            `import { log } from "@/test/logger";
import { bar } from "@test2/abc/bar";

const a = bar();

export default log();`,
        );
        writeFileSync(`${temporaryDirectoryPath}/src/test/logger.ts`, `export const log = () => console.log("test");`);
        writeFileSync(`${temporaryDirectoryPath}/src/test2/foo/bar.ts`, `export const bar = () => console.log("bar");`);
        createPackageJson(temporaryDirectoryPath, {
            main: "./dist/index.cjs",
            module: "./dist/index.mjs",
            type: "commonjs",
        });
        createTsConfig(temporaryDirectoryPath, { compilerOptions: { rootDir: "./src" } });
        await createPackemConfig(temporaryDirectoryPath, {
            alias: {
                "@": resolve(temporaryDirectoryPath, "src"),
                "@test2/abc": resolve(temporaryDirectoryPath, "src", "test2", "foo"),
            },
        });

        const binProcess = await execPackemSync("build", [], {
            cwd: temporaryDirectoryPath,
        });

        await expect(streamToString(binProcess.stderr)).resolves.toBe("");
        expect(binProcess.exitCode).toBe(0);

        const mjsContent = readFileSync(`${temporaryDirectoryPath}/dist/index.mjs`);

        expect(mjsContent).toMatchSnapshot("mjs content");

        const cjsContent = readFileSync(`${temporaryDirectoryPath}/dist/index.cjs`);

        expect(cjsContent).toMatchSnapshot("cjs content");
    });

    it.each(["@", "#", "~"])("should trigger a error if a invalid alias (%s) was used", async (alias) => {
        expect.assertions(2);

        writeFileSync(
            `${temporaryDirectoryPath}/src/index.ts`,
            `import { log } from "${alias}/test/logger";

export default log();`,
        );
        writeFileSync(`${temporaryDirectoryPath}/src/test/logger.ts`, `export const log = () => console.log("test");`);
        createPackageJson(temporaryDirectoryPath, {
            main: "./dist/index.cjs",
            module: "./dist/index.mjs",
            type: "commonjs",
        });
        createTsConfig(temporaryDirectoryPath, { compilerOptions: { rootDir: "./src" } });
        await createPackemConfig(temporaryDirectoryPath, {
            alias: {
                [alias + "/"]: resolve(temporaryDirectoryPath, "src")
            },
        });

        const binProcess = await execPackemSync("build", [], {
            cwd: temporaryDirectoryPath,
            // This is needed to get the error
            reject: false,
        });

        expect(binProcess.exitCode).toBe(1);
        await expect(streamToString(binProcess.stderr)).resolves.toContain(`Error: Alias name "${alias}/" is invalid. Alias names should start with a letter or underscore and only contain letters, numbers, underscores, and dashes.`);
    });
});
