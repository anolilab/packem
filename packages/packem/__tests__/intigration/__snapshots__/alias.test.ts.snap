// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`packem alias > should not trigger a warning if alias option and original is used > cjs content 1`] = `
"'use strict';

var __defProp$1 = Object.defineProperty;
var __name$1 = (target, value) => __defProp$1(target, "name", { value, configurable: true });
const log = /* @__PURE__ */ __name$1(() => console.log("test"), "log");

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const bar = /* @__PURE__ */ __name(() => console.log("bar"), "bar");

bar();
const index = log();

module.exports = index;
"
`;

exports[`packem alias > should not trigger a warning if alias option and original is used > mjs content 1`] = `
"var __defProp$1 = Object.defineProperty;
var __name$1 = (target, value) => __defProp$1(target, "name", { value, configurable: true });
const log = /* @__PURE__ */ __name$1(() => console.log("test"), "log");

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const bar = /* @__PURE__ */ __name(() => console.log("bar"), "bar");

bar();
const index = log();

export { index as default };
"
`;

exports[`packem alias > should not trigger a warning if alias option is used > cjs content 1`] = `
"'use strict';

var __defProp$1 = Object.defineProperty;
var __name$1 = (target, value) => __defProp$1(target, "name", { value, configurable: true });
const log = /* @__PURE__ */ __name$1(() => console.log("test"), "log");

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const bar = /* @__PURE__ */ __name(() => console.log("bar"), "bar");

bar();
const index = log();

module.exports = index;
"
`;

exports[`packem alias > should not trigger a warning if alias option is used > mjs content 1`] = `
"var __defProp$1 = Object.defineProperty;
var __name$1 = (target, value) => __defProp$1(target, "name", { value, configurable: true });
const log = /* @__PURE__ */ __name$1(() => console.log("test"), "log");

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const bar = /* @__PURE__ */ __name(() => console.log("bar"), "bar");

bar();
const index = log();

export { index as default };
"
`;
